### PR TITLE
hw was implemented

### PR DESCRIPTION
**Pull request by Mandybur800**

### CODE REVIEW

File: src/main/java/ua/com/conductor/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            }
+        }
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> getByEmailQuery = session.createQuery("FROM User WHERE email = :email",
+                    User.class);
+            getByEmailQuery.setParameter("email", email);
```
Comment (IvanLuchkin) [inline src/main/java/ua/com/conductor/dao/impl/UserDaoImpl.java]: Can be written in one line, but up to you

File: src/main/java/ua/com/conductor/model/User.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "user")
```
Comment (IvanLuchkin) [inline src/main/java/ua/com/conductor/model/User.java]: It's better to use plural for table names. Moreover, such name for the table may conflict with internal MySQL table for users

File: src/main/java/ua/com/conductor/service/impl/AuthenticationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        throw new AuthenticationException("Incorrect login or password");
+    }
+
+    @Override
+    public User register(String email, String password) {
+        User user = new User();
+        user.setEmail(email);
+        byte[] salt = HashUtil.getSalt();
+        String hashPassword = HashUtil.hashPassword(password, salt);
```
Comment (IvanLuchkin) [inline src/main/java/ua/com/conductor/service/impl/AuthenticationServiceImpl.java]:
```suggestion
        String hashedPassword = HashUtil.hashPassword(password, salt);
```

**Review by IvanLuchkin — CHANGES_REQUESTED** [review]

**Review by IvanLuchkin — APPROVED** [review]

File: src/main/java/ua/com/conductor/dao/impl/UserDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            return session.createQuery("FROM User WHERE email = :email",
+                    User.class).setParameter("email", email).uniqueResultOptional();
```
Comment (dmytro-hryhoruk) [inline src/main/java/ua/com/conductor/dao/impl/UserDaoImpl.java]:
```suggestion
            return session
                  .createQuery("FROM User WHERE email = :email", User.class)
                  .setParameter("email", email)
                  .uniqueResultOptional();
```

File: src/main/java/ua/com/conductor/service/impl/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> optionalUser = userService.findByEmail(email);
+        if (optionalUser.isEmpty()) {
+            throw new AuthenticationException("Incorrect login");
```
Comment (dmytro-hryhoruk) [inline src/main/java/ua/com/conductor/service/impl/AuthenticationServiceImpl.java]: let's not make the life of some hackers easier) 
Let's not refactor this method the way so we are throwing AuthenticationException only once and only with the message: 
"Incorrect login or password"; for example you can user optionalUser.isPresent() with &&

File: src/main/java/ua/com/conductor/service/impl/AuthenticationServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    @Override
+    public User register(String email, String password) {
+        User user = new User();
+        user.setEmail(email);
+        byte[] salt = HashUtil.getSalt();
+        String hashedPassword = HashUtil.hashPassword(password, salt);
+        user.setSalt(salt);
+        user.setPassword(hashedPassword);
+        userService.add(user);
+        return user;
```
Comment (dmytro-hryhoruk) [inline src/main/java/ua/com/conductor/service/impl/AuthenticationServiceImpl.java]: checklist

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**